### ISSUE TITLE
Providing Supply and Demand node in the OpenStudio Application for WaterHeater:Mixed

### ISSUE THREAD

**Issue by BehnamMohseniGharyehsafa** [openstudiocoalition/OpenStudioApplication#772]
Enhancement Request: @DavidGoldwasser 
--------------

# General summary of the feature:

Providing Supply and Demand node in the OpenStudio Application for WaterHeater:Mixed

# Detailed Description:

The WaterHeater:Mixed object includes the supply and demand node features in the IDF and EnergyPlus text modes, **but it lacks the capability to explicitly define these nodes within the OpenStudio Application.** Introducing this feature to OpenStudio would significantly streamline the integration of detailed HVAC systems, particularly in complex setups like those involving thermal storage tanks for energy flexibility. This enhancement could help reduce errors, especially in cases with multiple zones or intricate heating strategies, aligning with the broader goal of improving energy management systems.

# Possible Implementation:

IronBug, a plugin from LadyBug Tools, leverages OpenStudio features to streamline the implementation of detailed HVAC systems. Its methodology simplifies the design of HVAC systems, even in large-scale and complex case studies. Many of these studies, which include systems such as boilers or heat pumps, incorporate water buffer storage tanks to provide additional hot water for space heating or service hot water. However, for buildings with numerous zones, correctly identifying and assigning node names becomes a significant challenge when implementing water storage systems.

Similarly, components like fluid-to-fluid heat exchangers in the OpenStudio application require explicit definition of supply and demand nodes. This feature is effectively integrated into IronBug, enabling users to easily specify these nodes. For example, the image below demonstrates the configuration of a fluid-to-fluid heat exchanger in IronBug, with clearly defined supply and demand sides.

![HeatExchanger_IronBug](https://github.com/user-attachments/assets/9fee4f31-5706-4a99-89ca-4eca40822f6b)

In contrast, the WaterHeater:Mixed object in OpenStudio lacks this capability, making it exceedingly difficult to define supply and demand nodes within both IronBug and OpenStudio. The only workaround involves manually hardcoding the IDF script—a process prone to errors, especially when assigning nodes recursively.

Adding this feature to the WaterHeater:Mixed object would greatly enhance the ease and accuracy of defining detailed HVAC systems.

Thank you for your time and assistance. I look forward to hearing from you soon.

Sincerely,
Behnam Mohseni Gharyehsafa
National University of Ireland Maynooth

**Comment by BehnamMohseniGharyehsafa** [comment]
@DavidGoldwasser Could the modifications and enhancements be applied to the OpenStudio SDK or application?

**Comment by macumber** [comment]
@BehnamMohseniGharyehsafa could you provide a screen shot of where you would like to add a WaterHeater:Mixed in the OpenStudio Application?  Is it possible you could do what you are looking to do with two plant loops coupled by a heat exchanger as shown below?

![image](https://github.com/user-attachments/assets/6784a764-c8a9-4f2f-952e-5901453f6dc7)

![image](https://github.com/user-attachments/assets/3fb9532a-0cce-401c-8227-29ca5fc9b9bc)

**Comment by BehnamMohseniGharyehsafa** [comment]
@macumber 
Thank you for your message

Pictures below are the configuration that I am interested in defining:

The primary loop:

![photo_2024-11-18_16-10-29](https://github.com/user-attachments/assets/dd96b05c-2ca3-464d-808f-762d487edab2)

And the secondary loop:

![photo_2024-11-18_16-11-36](https://github.com/user-attachments/assets/c7e8f5f3-977a-419d-b399-78142d03af70)

So the idea is that the EIR_HP runs for specific hours, let's say from 08:00 to 19:00. The distribution of the HP heat is different. From 08:00 to the end of 15:00, the generated HP heat is used to heat up the house via the baseboard (node 14 and node 13), and from 16:00 till 19:00, the HP generated heat is going to the heat exchanger to charge the water storage tank in the secondary loop (this should be done from node 19 and node 24 for the heat exchanger and node $\alpha$ and node $\beta$ for the WaterHeater:Mixed). So, the operation hours of the heat exchanger are from 16:00 to the end of 19:00.

The discharging hours of the WaterHeater:Mixed is from 20:00 to the end of 21:00, meaning that the baseboard included in the secondary loop, is in operation from 20:00 to the end of 21:00 and its heat is provided by the charged WaterHeater:Mixed. 

However, I cannot find a way to proceed with such a configuration and all I can get are the following pictures: 
Primary loop:

![image](https://github.com/user-attachments/assets/3b628814-19a3-40c5-97a4-13834285bc5f)


Secondary loop:

![image](https://github.com/user-attachments/assets/3c94d1d2-2565-4273-86f0-a0dc41f70eb1)

I have the OSM and the IDF attached in the below file. The part indicating the supply node for the WaterHeater:Mixed is empty, meaning that there is no charging for the WaterHeater:Mixed. 

[Behnam_GitHub_Question.zip](https://github.com/user-attachments/files/18113037/Behnam_GitHub_Question.zip)


Could you kindly let me know what I should do? 
How I can define the supply side nodes for the WaterHeater:Mixed? 
Would it be possible for you to make an OSM file with the configuration I drew in the first two pictures? 
Is it possible to update the OpenStudio app to have this configuration available in user-friendly mode?

Thank you very much in advance and I am looking forward to hearing from you.
 
Sincerely,
Behnam

**Comment by BehnamMohseniGharyehsafa** [comment]
@macumber 

Is there any chance to solve the challenge?

**Comment by macumber** [comment]
I think you want something like below, where the water heater is on the demand side of one plant loop but the supply side of another.  The trick here is to add the water heater from the Library on the first plant loop.  Then add the water heater from My Model to the second plant loop.  That will trigger some new nodes to appear which you can use to navigate between the two plant loops.  I hope that helps, if not you might want to post the question to unmethours.  As far as I can tell, the OS SDK and OS Application support the configuration you want.

![image](https://github.com/user-attachments/assets/bc449c42-5976-4f29-844a-3404ace4a7f9)

![image](https://github.com/user-attachments/assets/3a2a87aa-4a31-4071-a87d-38d7689d9855)

**Comment by BehnamMohseniGharyehsafa** [comment]
@macumber Thank you very much for your detailed explanation and the example you provided.

I drew inspiration from your example as well as an example shared with me by Joseph Robertson. I also analyzed other OpenStudio examples, including one featuring a WaterStorage system connected to a solar collector. Upon review, I realized that the SDK and OpenStudio already had the required functionality. I reported the challenge to the Ladybug Tools developers, and I am pleased to share that the system is now fully operational.

@DavidGoldwasser I sincerely appreciate your time, support, and assistance.
